File: Python/287.py
#1
class Solution:
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        size = len(nums)
        if size == 0:
            return 0
        i, j = nums[0], nums[0]
        while True:
            i = nums[i]
            j = nums[nums[j]]
            if i == j:
                break
        j = nums[0]
        while i != j:
            i = nums[i]
            j = nums[j]
        return i


#2
class Solution:
    def findDuplicate(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        for i in range(len(nums)):
            if nums[abs(nums[i]) - 1] < 0:
                return abs(nums[i])
            nums[abs(nums[i]) - 1] = -nums[abs(nums[i]) - 1]
        return 0

File: C++/287.cpp
// #1
class Solution {
public:
    int findDuplicate(vector<int> &nums) {
        auto size = nums.size();
        if (size == 0)
            return 0;
        int fast = nums[0], slow = nums[0];
        do {
            fast = nums[nums[fast]];
            slow = nums[slow];
        } while (fast != slow);
        fast = nums[0];
        while (fast != slow) {
            slow = nums[slow];
            fast = nums[fast];
        }
        return slow;
    }
};


// #2
class Solution {
public:
    int findDuplicate(vector<int> &nums) {
        for (int i = 0; i < nums.size(); ++i) {
            int &cur = nums[abs(nums[i]) - 1];
            if (cur < 0)
                return abs(nums[i]);
            cur = -cur;
        }
        return 0;
    }
};
